name-registry/src/main.sw

```diff
@@ -0,0 +1,165 @@
+use interface::NameRegistry;
+use std::{
+    block::timestamp,
+    chain::auth::msg_sender,
+    constants::{
+        BASE_ASSET_ID,
+    },
+    context::call_frames::msg_asset_id,
+    context::msg_amount,
+    logging::log,
+    option::Option,
+    result::Result,
+    revert::{
+        require,
+        revert,
+    },
+    storage::StorageMap,
+};
+
+storage {
+    /// A mapping of names to an option of records, with a none representing an unregistered name
+    names: StorageMap<str[8], Option<Record>> = StorageMap {},
+}
+
+/// The amount to charge per hundred seconds per name
+const PRICE_PER_HUNDRED: u64 = 1;
+
+impl NameRegistry for Contract {
+    #[storage(read)]
+    fn expiry(name: str[8]) -> u64 {
+        match storage.names.get(name) {
+            Option::Some(record) => {
+                record.expiry
+            },
+            None => {
+                log(Errors::NameNotRegistered);
+                revert(0)
+            }
+        }
+    }
+
+    #[storage(read, write)]
+    fn extend(name: str[8], duration: u64) {
+        require(storage.names.get(name).is_some(), Errors::NameNotRegistered);
+        require(duration / 100 * PRICE_PER_HUNDRED <= msg_amount(), Errors::InsufficientPayment);
+        require(msg_asset_id() == BASE_ASSET_ID, Errors::WrongAssetSent);
+
+        let old_record = storage.names.get(name).unwrap();
+        let new_record = Record {
+            expiry: old_record.expiry + duration,
+            identity: old_record.identity,
+            owner: old_record.owner,
+        };
+
+        storage.names.insert(name, Option::Some(new_record));
+
+        log(RegistrationExtended {
+            duration,
+            name,
+            new_expiry: new_record.expiry,
+        });
+    }
+
+    #[storage(read)]
+    fn identity(name: str[8]) -> Identity {
+        match storage.names.get(name) {
+            Option::Some(record) => {
+                record.identity
+            },
+            None => {
+                log(Errors::NameNotRegistered);
+                revert(0)
+            }
+        }
+    }
+
+    #[storage(read)]
+    fn owner(name: str[8]) -> Identity {
+        match storage.names.get(name) {
+            Option::Some(record) => {
+                record.owner
+            },
+            None => {
+                log(Errors::NameNotRegistered);
+                revert(0)
+            }
+        }
+    }
+
+    #[storage(read, write)]
+    fn register(name: str[8], duration: u64) {
+        if storage.names.get(name).is_some() {
+            let record = storage.names.get(name).unwrap();
+            require(timestamp() > record.expiry, Errors::NameNotExpired);
+        }
+
+        require(duration / 100 * PRICE_PER_HUNDRED <= msg_amount(), Errors::InsufficientPayment);
+        require(msg_asset_id() == BASE_ASSET_ID, Errors::WrongAssetSent);
+
+        let record = Record {
+            expiry: timestamp() + duration,
+            identity: msg_sender().unwrap(),
```

Comment:
Just a thought but do you want to include the ability to resolve to a different `identity` than the one who is registering it? This would be passed into the function itself in that case and the front end would handle it.

Reply:
Can you elaborate? I dont get what you meant

Reply:
Your function definition would be `fn register(duration: u64, resolver: Identity, name: str[8]);` and you set `resolver` to the `identity` field of your `Record` struct. This would eliminate the user's need to make an additional call to set which `Identity` the registry should resolve to and save them some gas. I've done this personally in the past with ENS where I purchased an ENS from one wallet and set it to point to another

Reply:
The user would never have to make 2 transactions, the two abi calls would be made in the same script. But yeah sure I can change it to allow you to set the identity in the same call